source/logrepl/combined.go

```diff
@@ -0,0 +1,214 @@
+// Copyright © 2022 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package logrepl
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/conduitio/conduit-connector-postgres/source/position"
+	"github.com/conduitio/conduit-connector-postgres/source/snapshot"
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	"github.com/jackc/pgx/v5/pgxpool"
+)
+
+type iterator interface {
+	Next(context.Context) (sdk.Record, error)
+	Ack(context.Context, sdk.Position) error
+	Teardown(context.Context) error
+}
+
+type CombinedIterator struct {
+	conf             Config
+	pool             *pgxpool.Pool
+	cdcIterator      *CDCIterator
+	snapshotIterator *snapshot.Iterator
+
+	activeIterator iterator
+}
+
+type Config struct {
+	Position        sdk.Position
+	SlotName        string
+	PublicationName string
+	Tables          []string
+	TableKeys       map[string]string
+	WithSnapshot    bool
+}
+
+// NewCombinedIterator will initialize and start the Snapshot and CDC iterators.
+// Failure to parse the position or validate the config will return an error.
+func NewCombinedIterator(ctx context.Context, pool *pgxpool.Pool, conf Config) (*CombinedIterator, error) {
+	pos, err := position.ParseSDKPosition(conf.Position)
+	if err != nil {
+		sdk.Logger(ctx).Debug().
+			Err(err).
+			Msgf("failed to parse position: %s", string(conf.Position))
+
+		return nil, fmt.Errorf("failed to create logrepl iterator: %w", err)
+	}
+
+	if err := conf.Validate(); err != nil {
+		return nil, fmt.Errorf("failed to validate longrepl config: %w", err)
+	}
+
+	c := &CombinedIterator{
+		conf: conf,
+		pool: pool,
+	}
+
+	// Initialize the CDC iterator. If snapshotting is disabled or position is of
+	// CDC type, the iterator will be started as well.
+	if err := c.initCDCIterator(ctx, pos); err != nil {
+		return nil, err
+	}
+
+	// Initialize the snapshot iterator when snapshotting is enabled and not completed.
+	if conf.WithSnapshot && pos.Type != position.TypeCDC {
+		if err := c.initSnapshotIterator(ctx); err != nil {
+			return nil, err
+		}
+	}
```

Comment:
One more suggestion to make this easier to read:
1. I'd move the condition into `initSnapshotIterator`, let the function decide when a snapshot iterator is needed.
2. I'd add an explicit step _after_ initializing both iterators, where we choose the active iterator. It can be as simple as checking if the snapshot iterator exists.

```suggestion
	if err := c.initSnapshotIterator(ctx, pos); err != nil {
		return nil, fmt.Errorf("failed to initialize snapshot iterator: %w", err)
	}
	
	// choose active iterator
	switch {
	case c.snapshotIterator != nil:
		c.activeIterator = c.snapshotIterator
	default:
		c.activeIterator = c.cdcIterator
		if err := c.cdcIterator.Start(ctx); err != nil {
			return fmt.Errorf("failed to start CDC iterator: %w", err)
		}
	}
```

Reply:
yea good point, originally, I had the condition inside the init but decide to move it to be explicit.